File: tests/manual/algo_bench/lua/pathfind.lua
-- Grid pathfinding: A* and Dijkstra. Port of py/pathfind.py (0-based ids kept;
-- arrays are 1-based so index = id + 1).
local W = 128
local H = 128
local N = W * H
local BIG = 1000000000
local QUERIES = 40
local WALL_PCT = 30

local seed = 42
local function rnd()
  seed = (seed * 16807) % 2147483647
  return seed
end

local function iabs(x)
  if x < 0 then return -x end
  return x
end

local Heap = {}
Heap.__index = Heap
function Heap.new()
  return setmetatable({keys = {}, vals = {}, n = 0}, Heap)
end
function Heap:push(key, val)
  local keys, vals = self.keys, self.vals
  local i = self.n
  self.n = i + 1
  while i > 0 do
    local p = (i - 1) // 2
    if keys[p + 1] <= key then break end
    keys[i + 1] = keys[p + 1]
    vals[i + 1] = vals[p + 1]
    i = p
  end
  keys[i + 1] = key
  vals[i + 1] = val
end
function Heap:pop()
  local keys, vals = self.keys, self.vals
  local top = vals[1]
  local n = self.n - 1
  local last_key = keys[n + 1]
  local last_val = vals[n + 1]
  keys[n + 1] = nil
  vals[n + 1] = nil
  self.n = n
  if n > 0 then
    local i = 0
    while true do
      local l = 2 * i + 1
      if l >= n then break end
      local r = l + 1
      local m = l
      if r < n and keys[r + 1] < keys[l + 1] then m = r end
      if keys[m + 1] >= last_key then break end
      keys[i + 1] = keys[m + 1]
      vals[i + 1] = vals[m + 1]
      i = m
    end
    keys[i + 1] = last_key
    vals[i + 1] = last_val
  end
  return top
end

local function make_grid()
  local blocked = {}
  for i = 1, N do blocked[i] = 0 end
  for i = 1, N do
    if rnd() % 100 < WALL_PCT then blocked[i] = 1 end
  end
  return blocked
end

local function search(blocked, start, goal, use_heur)
  local gx = goal % W
  local gy = goal // W
  local dist = {}
  local closed = {}
  for i = 1, N do dist[i] = BIG; closed[i] = 0 end
  local heap = Heap.new()
  dist[start + 1] = 0
  heap:push(0, start)
  while heap.n > 0 do
    local cur = heap:pop()
    if closed[cur + 1] == 0 then
      closed[cur + 1] = 1
      if cur == goal then return dist[cur + 1] end
      local cx = cur % W
      local cy = cur // W
      local d = dist[cur + 1] + 1
      if cx > 0 then
        local nb = cur - 1
        if blocked[nb + 1] == 0 and d < dist[nb + 1] then
          dist[nb + 1] = d
          if use_heur then heap:push(d + iabs(cx - 1 - gx) + iabs(cy - gy), nb) else heap:push(d, nb) end
        end
      end
      if cx < W - 1 then
        local nb = cur + 1
        if blocked[nb + 1] == 0 and d < dist[nb + 1] then
          dist[nb + 1] = d
          if use_heur then heap:push(d + iabs(cx + 1 - gx) + iabs(cy - gy), nb) else heap:push(d, nb) end
        end
      end
      if cy > 0 then
        local nb = cur - W
        if blocked[nb + 1] == 0 and d < dist[nb + 1] then
          dist[nb + 1] = d
          if use_heur then heap:push(d + iabs(cx - gx) + iabs(cy - 1 - gy), nb) else heap:push(d, nb) end
        end
      end
      if cy < H - 1 then
        local nb = cur + W
        if blocked[nb + 1] == 0 and d < dist[nb + 1] then
          dist[nb + 1] = d
          if use_heur then heap:push(d + iabs(cx - gx) + iabs(cy + 1 - gy), nb) else heap:push(d, nb) end
        end
      end
    end
  end
  return -1
end

local function run(use_heur)
  seed = 42
  local blocked = make_grid()
  local total = 0
  local found = 0
  for q = 1, QUERIES do
    local s = rnd() % N
    local g = rnd() % N
    while blocked[s + 1] == 1 do s = rnd() % N end
    while blocked[g + 1] == 1 do g = rnd() % N end
    local r = search(blocked, s, g, use_heur)
    if r >= 0 then
      found = found + 1
      total = total + r
    end
  end
  return total * 1000 + found
end

local t0 = os.clock()
local a = run(true)
local t1 = os.clock()
local b = run(false)
local t2 = os.clock()
print("astar checksum", a)
print("dijkstra checksum", b)
print("astar", t1 - t0)
print("dijkstra", t2 - t1)

File: tests/manual/algo_bench/py/pathfind.py
# Grid pathfinding: A* and Dijkstra on a W x H grid with random walls.
# Same file runs under CPython and Zen (subset: no tuple-swap on subscripts,
# no abs/min/max builtins, lists built by comprehension).
import time

W = 128
H = 128
N = W * H
BIG = 1000000000
QUERIES = 40
WALL_PCT = 30

seed = 42
def rnd():
    global seed
    seed = (seed * 16807) % 2147483647
    return seed

def iabs(x):
    if x < 0:
        return -x
    return x

class Heap:
    # binary min-heap of (key, value) stored flat: keys[i], vals[i]
    def __init__(self):
        self.keys = []
        self.vals = []
        self.n = 0
    def push(self, key, val):
        keys = self.keys
        vals = self.vals
        i = self.n
        self.n = i + 1
        keys.append(key)
        vals.append(val)
        while i > 0:
            p = (i - 1) // 2
            if keys[p] <= key:
                break
            keys[i] = keys[p]
            vals[i] = vals[p]
            i = p
        keys[i] = key
        vals[i] = val
    def pop(self):
        # returns the value with the smallest key
        keys = self.keys
        vals = self.vals
        top = vals[0]
        last_key = keys.pop()
        last_val = vals.pop()
        n = self.n - 1
        self.n = n
        if n > 0:
            i = 0
            while True:
                l = 2 * i + 1
                if l >= n:
                    break
                r = l + 1
                m = l
                if r < n and keys[r] < keys[l]:
                    m = r
                if keys[m] >= last_key:
                    break
                keys[i] = keys[m]
                vals[i] = vals[m]
                i = m
            keys[i] = last_key
            vals[i] = last_val
        return top

def make_grid():
    blocked = [0 for i in range(N)]
    for i in range(N):
        if rnd() % 100 < WALL_PCT:
            blocked[i] = 1
    return blocked

def search(blocked, start, goal, use_heur):
    # returns path length in steps, or -1
    gx = goal % W
    gy = goal // W
    dist = [BIG for i in range(N)]
    closed = [0 for i in range(N)]
    heap = Heap()
    dist[start] = 0
    heap.push(0, start)
    expanded = 0
    while heap.n > 0:
        cur = heap.pop()
        if closed[cur]:
            continue
        closed[cur] = 1
        expanded += 1
        if cur == goal:
            return dist[cur]
        cx = cur % W
        cy = cur // W
        d = dist[cur] + 1
        # 4 neighbours, unrolled
        if cx > 0:
            nb = cur - 1
            if not blocked[nb] and d < dist[nb]:
                dist[nb] = d
                if use_heur:
                    heap.push(d + iabs(cx - 1 - gx) + iabs(cy - gy), nb)
                else:
                    heap.push(d, nb)
        if cx < W - 1:
            nb = cur + 1
            if not blocked[nb] and d < dist[nb]:
                dist[nb] = d
                if use_heur:
                    heap.push(d + iabs(cx + 1 - gx) + iabs(cy - gy), nb)
                else:
                    heap.push(d, nb)
        if cy > 0:
            nb = cur - W
            if not blocked[nb] and d < dist[nb]:
                dist[nb] = d
                if use_heur:
                    heap.push(d + iabs(cx - gx) + iabs(cy - 1 - gy), nb)
                else:
                    heap.push(d, nb)
        if cy < H - 1:
            nb = cur + W
            if not blocked[nb] and d < dist[nb]:
                dist[nb] = d
                if use_heur:
                    heap.push(d + iabs(cx - gx) + iabs(cy + 1 - gy), nb)
                else:
                    heap.push(d, nb)
    return -1

def run(use_heur):
    global seed
    seed = 42
    blocked = make_grid()
    total = 0
    found = 0
    for q in range(QUERIES):
        s = rnd() % N
        g = rnd() % N
        while blocked[s]:
            s = rnd() % N
        while blocked[g]:
            g = rnd() % N
        r = search(blocked, s, g, use_heur)
        if r >= 0:
            found += 1
            total += r
    return total * 1000 + found

t0 = time.perf_counter()
a = run(True)
t1 = time.perf_counter()
b = run(False)
t2 = time.perf_counter()
print("astar checksum", a)
print("dijkstra checksum", b)
print("astar", t1 - t0)
print("dijkstra", t2 - t1)
